text/0000-typed-context-injection.md

```diff
@@ -0,0 +1,833 @@
+
+```rust
+fn caller_1(cx: Cx<&mut System1, &mut System2, &System3>) {
+    // `cx` has a mutable reference to `System1` and (at least) an immutable reference to `System2`
+    // so this is fine. The unnecessary `&System3` reference is completely ignored.
+    consumer(cx);
+}
+
+fn caller_2() {
+    ...
+
+    // We can give the context in the order it shows up in the target type...
+    consumer(Cx::new((&mut system_1, &system_2)));
+
+    // ...but we could also provide the context in any other order and let coercion
+    // turn e.g. a `Cx<&mut System2, &mut System1>` into a `Cx<&mut System1, &System2>`.
+    consumer(Cx::new((&mut system_2, &mut system_1)));
+
+    // Heck, we could even pass the function additional completely unrelated components and coercion
+    // would still have our back!
+    //
+    // Coerces `Cx<&mut SomeOtherSystem, &mut System2, &mut System1>` into
+    // `Cx<&mut System1, &System2>`.
+    consumer(Cx::new((&mut some_other_system, &mut system_2, &mut system_1)));
+
+    // And if that flexibility wasn't enough, we can combine several contexts together to
+    // construct a full context like so:
+    let inherited_cx = Cx::new((&mut some_other_system, &mut system_2));
+
+    // Coerces `Cx<Cx<&mut SomeOtherSystem, &mut System2>, &mut System1>` into
+    // `Cx<&mut System1, &System2>`.
+    consumer(Cx::new((inherited_cx, &mut system_1)));
+}
+```
+
+Of course, this process is entirely borrow aware. When we coerce a context, only the components needed by the coercion are borrowed. All other components are completely untouched by the process.
+
+```rust
+fn example(cx: Cx<&mut CommonSystem, &mut System1, &mut System2, &mut OtherSystem>) {
+    let borrow_1: Cx<&CommonSystem, &mut System1> = cx;
+    let borrow_2: Cx<&CommonSystem, &mut System2> = cx;
+
+    // Both borrows can coexist.
+    let _ = (borrow_1, borrow_2);
+
+    // ...because the coercions desugar to the following:
+
+    // `CxRaw<T>` is an implementation detail we'll talk about in a bit. For right now,
+    // just think of it as a newtype around an arbitrary tuple type `T`.
+    let borrow_1: CxRaw<(&CommonSystem, &mut System1)> = CxRaw::new((
+        &*self.0.0,
+        &mut *self.0.1,
+    ));
+    let borrow_2: CxRaw<(&CommonSystem, &mut System2)> = CxRaw::new((
+        &*self.0.0,
+        &mut *self.0.2,
+    ));
+
+    let _ = (borrow_1, borrow_2);
+}
+```
+
+We can use these features to implement a much more convenient version of the `play_sound_at_path` method demonstrated in the [motivation section](#motivation).
+
+```rust
+pub trait AudioBufferCxExt {
+    fn play_sound_at_path(self, path: &str);
+}
+
+type AudioBufferCx<'a> = Cx<
+    &'a mut AudioBuffer,
+    &'a mut VirtualFileSystem,
+    &'a mut AudioCache,
+    &'a mut SoundVolumeOverrides,
+>;
+
+impl AudioBufferCxExt for AudioBufferCx<'_> {
```

Comment:
My point is exactly saying that `Cx` should be a trait not a type, because coercing between types arbitrarily like this is... you know, very un-rusty. Very dynamic and confusing to read. imo coercions must not be possible unless they are fully invertible, e.g. `Box<T>` can be coerced to `Box<dyn Trait>`, but it is still possible to convert it back to `Box<T>` if you know the `T`; converting `Cx<(&T, &U)>` to `&T` doesn't make sense because you cannot get back the `&U` part either way.

Reply:
I don't really know how to respond to it being "un-rusty" and "confusing to read." Having used the userland prototype in my own project to inject access tokens, I'd say it certainly helps reduce the messiness of context injection and makes the code easier to read.

What I can respond to, however, is the suggestion to make everything a trait. I disagree. In asking for a context inheriting a trait, you say "I want everything in a context that has `A`, `B`, and `C`." Compare that to a function taking `Cx<&A, &B, &C>`, which says "I want the parts of a context with `A`, `B`, and `C`. Do whatever you want with everything else."

Reply:
Lifetime subtyping is transitively equal. If `a = b` and `c = b`, `a == c` is true in a reasonable way. But extracting two different fields from the same context wouldn't make them equal. This completely distorts the meaning of the "equal" symbol.

Reply:
I'm not sure I understand the comment on transitivity. I also don't fully understand the consequences of not upholding some of the properties you list. Does this break the compiler or is this a user expectations problem?